lib/strings.nix

```diff
@@ -315,6 +315,23 @@ rec {
   */
   escapeNixString = s: escape ["$"] (builtins.toJSON s);
 
+  /* Encode string for safe use in URI segments
+
+     https://en.wikipedia.org/wiki/Percent-encoding
+
+     Type: string -> string
+
+     Example:
+       escapeUriSegment "foo/bar?baz=qux&quux=quuz#corge"
+       => "foo%2Fbar%3Fbaz%3Dqux%26quux%3Dquuz%23corge"
+  */
+  escapeUriSegment = arg:
```

Comment:
This sort of thing should not be done in Nix code because it's slow. You can also do escaping in the builder.